### ISSUE TITLE
IStream / OStream API

### ISSUE THREAD

**Issue by cessen** [cessen/openexr-rs#10]
Let's try to make OpenEXR play more nicely with the Rust I/O story.

The Rust standard library has the `Read`, `Write`, and `Seek` traits.  I'm thinking that we can change our `*InputFile` and `*OutputFile` types to take anything that implement the `Read+Seek` and `Write+Seek` respectively.

We can also still provide convenience functions for just giving a file path.

Finally, we have a way to read OpenEXR data from memory, but not a way to write to memory.  Reading from memory takes a slice, which makes sense.  But writing will require the ability to grow the data.  So maybe writing to memory takes a `&mut Vec`?

**Comment by cessen** [comment]
@Ralith, would love your input on this!

**Comment by cessen** [comment]
So, looking at the docs for `Vec`, apparently it implements `Write` already for `Vec<u8>`.  Moreover, it can be wrapped in `Cursor` to provide seek functionality.  So if we support `Write+Seek` then we get writing to a Vec more-or-less for free.

Further, `Read` is implemented on `&[u8]`, so we could also get rid of the specific methods for reading from memory as well.

**Comment by Ralith** [comment]
This should be doable with some FFI gymnastics, and seems like a nice clean interface. I bet we could even propagate the Rust error types by throwing a custom exception in the C++ bindings for `Read`/`Write`+`Seek` and catching it in a special case in the EXR bindings.

I'm not totally sure of the use cases for this level of abstraction on the read side, but the write side is certainly justified and if nothing else symmetry is nice. My one reservation is that OpenEXR appears to have optimizations to avoid gratuitously copying data that's already in-memory, so we may want to keep the hardcoded support for reading directly from memory, rather than subsuming it into the generic interface.

**Comment by cessen** [comment]
> My one reservation is that OpenEXR appears to have optimizations to avoid gratuitously copying data that's already in-memory, so we may want to keep the hardcoded support for reading directly from memory, rather than subsuming it into the generic interface.

Ah, yeah, that's a fair point.  Okay, let's keep that, and we'll add general `Read+Seek` and `Write+Seek` support to handle the general case.

Are you interested in taking this on?

**Comment by Ralith** [comment]
I'm not sure I can expect to get to it in the next few days, but I can probably give it a go this weekend at the latest if you don't get there first. Exposing the Rust traits to C++ seems like an interesting exercise.

**Comment by cessen** [comment]
Ah, no worries.  I think I'll take a crack at it then.  I'm sick, and so have a bit of time due to not being at work.  I'll make a branch for it.

**Comment by cessen** [comment]
I've made a first attempt at it for output:
https://github.com/cessen/openexr-rs/tree/iostream

However, it's not working properly.  After calling `CEXR_OutputFile_from_stream()` the header in the returned `CEXR_OutputFile` appears to be wrong (reports the wrong resolution), and eventually things segfault.

I've spent a bit of time trying to track it down, but that's as far as I've gotten.  I remember now why I love Rust so much... memory safety bugs are so annoying.

Anyway, feel free to either take a look at what I've got so far, or make an attempt of your own (or both).

**Comment by cessen** [comment]
I simplified the implementation somewhat, but still getting the same error.

I was initially thinking that the calls to the writing Rust code were carrying incorrect addresses or something, and so the Rust code was then overwriting memory somewhere.  But that doesn't appear to be the case: I created a `DummyWriter` class that does zero writing to any memory, it just prints what the caller is asking it to do, and even with that class the same problem is still showing up.

And it's really consistent, too.  The header always ends up thinking it has a resolution of 0x63.  Also, all the channel descriptions are missing from the header.

**Comment by cessen** [comment]
Fixed it!

The issue was... surprising to me.  Maybe I've been away from C++ for too long.  There are two constructors for the C++ OutputFile class:

```c++
OutputFile (const char fileName[], const Header &header,
    int numThreads = globalThreadCount());

OutputFile (OPENEXR_IMF_INTERNAL_NAMESPACE::OStream &os,
    const Header &header, int numThreads = globalThreadCount());
```

I forgot to pass the header reference, so I guess `numThreads` was getting the default value and the thread count was being passed to `header`.  It's not at all clear to me why that compiles... I didn't recall C++ being *that* weak on types.  But that's what it was.

Now that I made that fix, everything seems to be working correctly.

@Ralith: I'd like to have your eyes on these changes, so I'll make a pull request so you can easily review it.  But no rush!

**Comment by cessen** [comment]
@Ralith

Okay, now that we've got output working (PR #11), let's do input!  This should be much easier, since it's pretty much exactly the same thing, and we already worked out how to do it for output.

However, if we want to keep the optimized reading from a slice, I'm worried we might need to do similar shenanigans as I did when trying to preserve the the original `ScanlineOutputFile::new()`.  If we have to do that, I won't be torn up or anything.  But do you have any ideas how to do things... cleaner?

I'll get started regardless.

**Comment by cessen** [comment]
Implemented in PRs #11, #13, and #15.
